homeassistant/components/stream/__init__.py

```diff
@@ -0,0 +1,156 @@
+"""
+Provide functionality to stream video source.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/stream/
+"""
+import logging
+import threading
+
+import voluptuous as vol
+
+from homeassistant.auth.util import generate_secret
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.loader import bind_hass
+
+from .const import (
+    DOMAIN, CONF_KEEPALIVE, ATTR_STREAMS, ATTR_ENDPOINTS)
+from .worker import stream_worker
+from .hls import async_setup_hls
+
+REQUIREMENTS = ['av==6.1.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['http']
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_KEEPALIVE, default=False): cv.boolean,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+# pylint: disable=dangerous-default-value
+@bind_hass
+async def async_request_stream(hass, stream_source, options={}, preload=False):
+    """Set up stream with token."""
+    if DOMAIN not in hass.config.components:
+        raise HomeAssistantError("Stream is not configured.")
+
+    if not stream_source:
+        raise HomeAssistantError("Invalid stream source.")
+
+    keepalive = hass.data[DOMAIN][CONF_KEEPALIVE]
+    try:
+        streams = hass.data[DOMAIN][ATTR_STREAMS]
+        stream = streams.get(stream_source)
+        if not stream:
+            stream = Stream(hass, stream_source,
+                            options=options, keepalive=keepalive)
+            hass.data[DOMAIN][ATTR_STREAMS][stream_source] = stream
+        if not preload and not stream.access_token:
```

Comment:
Wait, what is the idea of having an access token be set via the options? That doesn't seem like something we should want to outsource.

Reply:
The preload is in here mostly for the test right now, but can be used in a future commit to load the stream and add a provider on Home Assistant load, without starting the stream.  The problem I ran into with the test is the thread would complete before the HLS provider could be added to the stream since It's just quickly demuxing a 3 second test video clip.  The segments wouldn't exist at that point.

The reason I have the access token check here is around your previous comment regarding the request_stream being called multiple times (cast and frontend) and not wanting to invalidate the previous token, so we just re-use it here.  The token is not set by the options that will be exposed to the user, just combining similar code into a multi-purpose method.

Reply:
Would it make sense that the stream constructor will generate the access token ?

Reply:
it feels a bit weird to do things for a test. Can't we make the video longer?

Reply:
The access token can not be generated by the constructor because it needs to be re-generated after it expires, and we never get rid of the Stream instance after we expire the stream.  We just remove providers, clear the token, and stop the thread.

Reply:
I removed the preload option and the function giving direct access to the stream in favor of setting the stream directly on `hass.data` in the tests.